packages/modal/src/ux-modal-service.ts

```diff
@@ -0,0 +1,403 @@
+  };
+  theme?: UxModalTheme;
+  keyboard?: UxModalKeybord;
+  restoreFocus?: (lastActiveElement: HTMLElement) => void;
+  openingCallback?: (contentWrapperElement?: HTMLElement, overlayElement?: HTMLElement) => void;
+  host?: HTMLElement;
+  dismiss?: () => void;
+  ok?: (event: Event) => void;
+}
+
+interface UxModalLayer {
+  bindingContext?: UxModalBindingContext;
+  modal: UxModal;
+}
+
+@inject(TemplatingEngine, CompositionEngine, ViewResources, EventAggregator, UxDefaultModalConfiguration)
+export class UxModalService {
+
+  public startingZIndex: number = 200;
+  public modalLayers: Array<UxModalLayer> = [];
+  public modalIndex: number = 0;
+
+  private handleDocumentClick: (event: Event) => void;
+
+  constructor(
+    private templatingEngine: TemplatingEngine,
+    private compositionEngine: CompositionEngine,
+    private viewResources: ViewResources,
+    private eventAggregator: EventAggregator,
+    private defaultConfig: UxDefaultModalConfiguration) {
+    this.handleDocumentClick = (event: Event) => {
+      // the purpose of this handler is to close all modals that are
+      // - not locked
+      // - have outsideDismiss === true
+      // - placed above the last locked layer
+      // - and if the click is outside the modal
+      let concernedLayers: Array<UxModalLayer> = [];
+      for (const layer of this.modalLayers) {
+        if (layer.modal.lock || !layer.modal.outsideDismiss) {
+          concernedLayers = [];
+          continue; // we ignore locks
+        }
+        concernedLayers.push(layer);
+      }
+      // now we have all the layers above the last locked in the concernedLayers array
+      // let's check if the click is outside of any
+      layerloop: for (const layer of concernedLayers) {
+        const modalContentElement = layer.modal.contentElement;
+        for (const element of (event as any).composedPath() ) {
+          if (element === modalContentElement) {
+            continue layerloop; // click is on the modal, do not hide
+          }
+        }
+        layer.modal.dismiss();
+      }
+      return true; // this allow normal behvior with this click for other purposes
+    };
+  }
+
+  public addLayer(modal: UxModal, bindingContext: UxModalBindingContext) {
+    const layerCount = this.modalLayers.push({
+      bindingContext,
+      modal
+    });
+    if (layerCount === 1) {
+      this.setListener();
+    }
+  }
+
+  public getLayer(modal: UxModal): UxModalLayer | undefined {
+    const index = this.modalLayers.map(i => i.modal).indexOf(modal);
+    if (index !== -1) {
+      return this.modalLayers[index];
+    }
+    return undefined;
+  }
+
+  public removeLayer(modal: UxModal) {
+    const index = this.modalLayers.map(i => i.modal).indexOf(modal);
+    if (index !== -1) {
+      this.modalLayers.splice(index, 1);
+    }
+    if (this.modalLayers.length === 0) {
+      this.removeListener();
+    }
+  }
+
+  private setListener() {
+    window.addEventListener('keyup', this);
```

Comment:
why is key on window and click on document? If we do both of them in either window or document, I think we can use `this` as listener for both and switch `event.type` to call the corresponding handler instead

Reply:
Good idea https://github.com/aurelia/ux/pull/255/commits/08a48f25047313c259dc23f54428d4fbcd21d008